activerecord/lib/active_record/attribute_methods.rb

```diff
@@ -463,6 +464,20 @@ def attributes_for_create(attribute_names)
         end
       end
 
+      # Mask sensitive data and format value for inspection.
+      # +nil+ is not filtered even if it matches +filter_parameter+.
+      def attributes_for_inspect(filter_parameter)
+        filter = ActiveSupport::ParameterFilter.new(filter_parameter)
+        hash = self.class.attribute_names.inject({}) do |o, attr_name|
+          if has_attribute?(attr_name)
+            value = read_attribute(attr_name)
+            o[attr_name] = value.nil? ? value : attribute_for_inspect(attr_name)
+          end
+          o
+        end
+        filter.filter(hash, filter_nil: false).transform_values! { |v| v.nil? ? v.inspect : v }
```

Comment:
This usage suggests that a `ParameterFilter#filter_param(key, value)` API would be desirable so we don't have to preprocess, filter, then postprocess an attributes hash. Instead we could enumerate `attributes` directly:
```ruby
def attributes_for_inspect
  attributes.each_with_object({}) do |(attr_name, value), filtered|
    filtered[attr_name] = attribute_value_for_inspect(self.class.inspect_filter.filter_param(attr_name, value)))
  end
end

# change implementation to delegate to new value-formatter method…
def attribute_for_inspect(attr_name)
  attribute_value_for_inspect(read_attribute(attr_name))
end

# Extract implementation from current attribute_for_inspect
def attribute_value_for_inspect(value)
  …
end
```

Reply:
I also considered that path, but gave up since proc filters with arity of 3 need an entire hash.
https://github.com/rails/rails/blob/99c87ad2474d5c5b6e52ceac34c3cf9f9cb57f9f/activesupport/lib/active_support/parameter_filter.rb#L95

What should `filter_param(key, value)` API do if proc filter which takes three args is given?
Passing `nil` as third block parameter may be imcompatible since filters do not expect to get `nil`

Reply:
`filter_param` is new API so backward compatibility may not be severe than I thought.
I will add it.